packages/@atjson/source-html/src/converter/video-embeds.ts

```diff
@@ -0,0 +1,193 @@
+import Document, { Annotation, AnnotationConstructor } from "@atjson/document";
+import OffsetSource, {
+  getClosestAspectRatio,
+  VideoEmbed,
+  VideoURLs,
+  CaptionSource
+} from "@atjson/offset-annotations";
+import { Iframe, Anchor, Div } from "../annotations";
+
+function covers(a: Annotation<any>, b: Annotation<any>) {
+  return a.start < b.start && a.end > b.end;
+}
+
+function assert(value: any, message: string): asserts value {
+  if (!value) {
+    throw new Error(message);
+  }
+}
+
+function is<T extends AnnotationConstructor<any, any>>(
+  annotation: Annotation<any>,
+  Class: T
+): annotation is InstanceType<T> {
+  let AnnotationClass = annotation.getAnnotationConstructor();
+  return (
+    AnnotationClass.vendorPrefix === Class.vendorPrefix &&
+    annotation.type === Class.type
+  );
+}
+
+function isIframe(annotation: Annotation<any>) {
+  return is(annotation, Iframe) && annotation.attributes.src;
+}
+
+function isVimeoEmbed(annotation: Annotation<any>) {
+  return (
+    is(annotation, Iframe) &&
+    annotation.attributes.src?.indexOf("https://player.vimeo.com/video") === 0
+  );
+}
+
+function isBrightcoveEmbed(annotation: Annotation<any>) {
+  return (
+    is(annotation, Iframe) &&
+    annotation.attributes.src?.indexOf("https://players.brightcove.net") === 0
+  );
+}
+
+function isBrightcoveWrapper(annotation: Annotation<any>) {
+  return (
+    is(annotation, Div) && annotation.attributes.style?.match(/padding\-top/)
+  );
+}
+
+function isVimeoLink(annotation: Annotation<any>) {
+  return (
+    is(annotation, Anchor) &&
+    annotation.attributes.href?.indexOf("https://vimeo.com")
+  );
+}
+
+function getSize(annotation: Iframe, name: "width" | "height") {
+  let dimension = annotation.attributes[name];
+  if (dimension) {
+    return parseInt(dimension, 10);
+  }
+  return undefined;
+}
+
+export default function(doc: Document) {
+  doc
+    .where(isVimeoEmbed)
+    .as("video")
+    .outerJoin(
+      doc.where({ type: "-html-p" }).as("paragraph"),
+      function findPossibleVimeoCaptions(embed, paragraph) {
+        return (
+          paragraph.start === embed.end || paragraph.start === embed.end + 1
+        );
+      }
+    )
+    .outerJoin(doc.where(isVimeoLink).as("links"), function vimeoCaptionJoin(
+      { paragraph },
+      link
+    ) {
+      return covers(paragraph[0], link);
+    })
+    .update(function convertVimeoVideos({ video, paragraph }) {
+      let url = VideoURLs.identify(new URL(video.attributes.src));
+      assert(
+        url,
+        `The Vimeo embed ${video.attributes.src} was definitely defined in our queries, but was not identified.`
+      );
+
+      let caption: CaptionSource | undefined;
+      if (paragraph.length) {
```

Comment:
i think you need to check if you have `links` here, as since you did an outer join in the previous step, `paragraphs` will still be non-empty even if the next paragraph contains no links